### ISSUE TITLE
Nuget Publisher iterates all files while applying filters.

### ISSUE THREAD

**Issue by EricABC** [Microsoft/vsts-tasks#3089]
Despite using globing patterns that would exclude large folders such as node_modules, every file seems to be iterated.  This is adding 10 min to a single publish action.

**Comment by bryanmacfarlane** [comment]
zack - work with eric - he's been focusing a lot on our find match lib code,

**Comment by zjrunner** [comment]
Another user hit a related issue that we fixed a while back (but rolling out now) -- essentially there was a truncation flaw where a specific method stripped more than we expected because our string didn't end with a directory separator.  This led to the search happening from the parent of the desired location which could sometimes include a much broader search than expected.  I had been lumping this issue in with that, but i'll check with eric on the build team again to see if we're missing something else.

**Comment by ericsciple** [comment]
No there is nothing to exclude lower level directories from being traversed today. Multiple include patterns would be a workaround. Also a solid state drive would likely make the additional impact negligible.

It's feasible to come up with a solution in the code, but I don't know how much effort it would be. The findMatch function in the task lib would be the most appropriate place to handle something like that. It already calls into minimatch and performs some analysis based on the parsed pattern object (broken up into sub-objects for each directory segment of each pattern). Also the patterns are applied in order, so you can have interleaved include/exclude. So some additional care would need to be taken based on the order.

**Comment by zjrunner** [comment]
Ok, that's what I was afraid of.  @EricABC can you double check your perf after our rollout next week (verison of the nuget task in your logs will be 0.2.27 or higher)?  See if the other optimizations made a difference for you first.  If not enough, can you see if you can be more selective in your search paths (is there a set of paths you could use that avoids the node_modules folder, or could you have your build place your packages into the build staging folder and publish from there?  If none of these are acceptable, please give us more information on the layout that makes the above difficult and we'll consider if we have more options when we do our next big update to the tasks.

**Comment by EricABC** [comment]
Our builds are actually running on some pretty beefy private agent Azure VMs with premium SSD (including OS drive).  Moving folder structures around for us isn't really an option as our devops is a bit complex, with layers of tooling to support angular 2 offline compilation across numerous library projects.  Exclusions won't really affect this line [here](https://github.com/Microsoft/vsts-tasks/blob/b5ac481d42842d96ce1cc51a1951652b9b85413d/Tasks/Common/nuget-task-common/Utility.ts#L94).  I had tried some very specific paths and it did not prevent the scan, although perhaps the wildcard change since would give me what I need.  There is always the option to add intermediate steps to copy files, but then I might as well have cmdline scripted the publish command at that point.

If you are already using the defacto glob library, why iterate all files on that line?  You could really just reduce the entire pattern input to a single glob pattern, which already supports compounding multiple globs.  Let glob do all the hard work of pruning, get the result back, and then do any further checking on that list.

The work around is really something we should have already done anyway: reduce build and release process to a pre and post build script.  Then we can share it across all our team projects instead of duplicating them.  But this fix would be useful even for very simple default Visual Studio web projects.

**Comment by zjrunner** [comment]
Yes, the find is still expensive, but in the previous case the find was happening a level further up the tree which can be *really* expensive in some cases (iterating historical build attempts or other things for some folks depending on what your search paths look like).  I'll have to see why a more standard glob usage with compounding isn't done, but I'm fairly certain there is a back compat consideration in here -- the tasks exist from before npm tasks were an option and it was all powershell and in moving forward we had to mimic the search style there or break people on the auto-upgrade.  Now that major versioning is a thing we have some cleaner ways to do breaking changes, but it still requires thought/support/education and is work for those we would break so we have to tread lightly at times.

**Comment by EricABC** [comment]
No worries, I'm just living with the delay for now, and eventually we'll take the time to script it.  Thanks!

**Comment by Tasteful** [comment]
I think my slow publishing is related to this and regarding to the log it is starting to iterating over one parent directory level wrong. 

I fixed my problem with replacing the first `*` in the globbing pattern with the correct artifacts name and then everything was going fast again.

The new version 0.2.27 or higher, when will that be available to TFS 2017 onprem?

```
2017-03-01T16:58:14.1383091Z Set workingFolder to default: C:\SysData\agent2\tasks\NuGetPublisher\0.2.21
2017-03-01T16:58:14.5133120Z ##[debug]agent.workFolder=C:\SysData\agent2\_work
2017-03-01T16:58:14.5133120Z ##[debug]loading inputs and endpoints
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_F8085F49-91E9-4938-8B92-49D463DE034E
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_F8085F49-91E9-4938-8B92-49D463DE034E_PASSWORD
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_SCHEME_F8085F49-91E9-4938-8B92-49D463DE034E
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2017-03-01T16:58:14.5289378Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2017-03-01T16:58:14.5289378Z ##[debug]loading INPUT_CONNECTEDSERVICENAME
2017-03-01T16:58:14.5289378Z ##[debug]loading INPUT_NUGETFEEDTYPE
2017-03-01T16:58:14.5289378Z ##[debug]loading INPUT_NUGETVERSION
2017-03-01T16:58:14.5289378Z ##[debug]loading INPUT_SEARCHPATTERN
2017-03-01T16:58:14.5289378Z ##[debug]loading INPUT_VERBOSITY
2017-03-01T16:58:14.5289378Z ##[debug]loaded 11
2017-03-01T16:58:14.8101897Z ##[debug]check path : C:\SysData\agent2\tasks\NuGetPublisher\0.2.21\task.json
2017-03-01T16:58:14.8101897Z ##[debug]set resource file to: C:\SysData\agent2\tasks\NuGetPublisher\0.2.21\task.json
2017-03-01T16:58:14.8101897Z ##[debug]system.culture=en-US
2017-03-01T16:58:14.8258137Z ##[debug]check path : C:\Windows\system32\chcp.com
2017-03-01T16:58:14.8258137Z ##[debug]C:\Windows\system32\chcp.com=C:\Windows\system32\chcp.com
2017-03-01T16:58:14.8258137Z ##[debug]C:\Windows\system32\chcp.com arg: ["65001"]
2017-03-01T16:58:14.8258137Z ##[debug]exec tool: C:\Windows\system32\chcp.com
2017-03-01T16:58:14.8258137Z ##[debug]Arguments:
2017-03-01T16:58:14.8258137Z ##[debug]   65001
2017-03-01T16:58:14.8258137Z [command]C:\Windows\system32\chcp.com 65001
2017-03-01T16:58:14.8414410Z Active code page: 65001
2017-03-01T16:58:14.8414410Z ##[debug]searchPattern=C:\SysData\agent2\_work\8cb121257\*\nuget\*.nupkg
2017-03-01T16:58:14.8414410Z ##[debug]System.DefaultWorkingDirectory=C:\SysData\agent2\_work\8cb121257
2017-03-01T16:58:14.8414410Z ##[debug]pattern: C:\SysData\agent2\_work\8cb121257\*\nuget\*.nupkg, isNegative: false
2017-03-01T16:58:14.8414410Z ##[debug]Matching glob pattern: C:\SysData\agent2\_work\8cb121257\*\nuget\*.nupkg
2017-03-01T16:58:14.8414410Z ##[debug]Index of first wildcard: 34
2017-03-01T16:58:14.8414410Z ##[debug]find root dir: C:\SysData\agent2\_work
2017-03-01T17:00:56.2761602Z ##[debug]find C:\SysData\agent2\_work
2017-03-01T17:00:56.2761602Z ##[debug]416104 matches.
2017-03-01T17:01:14.7289051Z ##[debug]connectedServiceName=f8085f49-91e9-4938-8b92-49d463de034e
2017-03-01T17:01:14.7289051Z ##[debug]feedName=null
2017-03-01T17:01:14.7289051Z ##[debug]nuGetAdditionalArgs=null
```

@zjrunner @ericsciple

**Comment by alexmullans** [comment]
This is fixed by #4397.
